packages/kbn-analytics/src/application_usage_tracker.ts

```diff
@@ -0,0 +1,182 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Reporter } from './reporter';
+import { createApplicationUsageMetric, ApplicationUsageMetric } from './metrics';
+
+type TrackedApplication = Record<string, ApplicationUsageMetric>;
+interface ApplicationKey {
+  appId: string;
+  viewId: string;
+}
+
+export class ApplicationUsageTracker {
+  private trackedApplicationViews: TrackedApplication = {};
+  private reporter: Reporter;
+
+  private currentAppId?: string;
+  private currentApplicationKeys: ApplicationKey[] = [];
+
+  private beforeUnloadListener?: EventListener;
+  private onVisiblityChangeListener?: EventListener;
+  private onClickListener?: EventListener;
+
+  constructor(reporter: Reporter) {
+    this.reporter = reporter;
+  }
+
+  private createKey(appId: string, viewId: string): ApplicationKey {
+    return { appId, viewId };
+  }
+
+  private serializeKey({ appId, viewId }: ApplicationKey): string {
+    return `${appId}-${viewId}`;
+  }
+
+  private trackApplications(appKeys: ApplicationKey[]) {
+    for (const { appId, viewId } of appKeys.filter(Boolean)) {
+      const serializedKey = this.serializeKey({ appId, viewId });
+      if (typeof this.trackedApplicationViews[serializedKey] !== 'undefined') {
+        continue;
+      }
+      const metric = createApplicationUsageMetric(appId, viewId);
+      this.trackedApplicationViews[serializedKey] = metric;
+    }
+  }
+
+  private attachListeners() {
+    if (typeof window === 'undefined' || typeof document === 'undefined') {
+      return;
+    }
+
+    this.beforeUnloadListener = () => {
+      this.flushTrackedViews();
+    };
+
+    this.onVisiblityChangeListener = () => {
+      if (document.visibilityState === 'visible') {
+        this.resumeTracking();
+      } else if (document.visibilityState === 'hidden') {
+        this.pauseTracking();
+      }
+    };
+
+    this.onClickListener = () => {
+      this.updateActiveViewsClickCounters();
+    };
+
+    // Before leaving the page, make sure we store the current usage
+    window.addEventListener('beforeunload', this.beforeUnloadListener);
+
+    // Monitoring dashboards might be open in background and we are fine with that
+    // but we don't want to report hours if the user goes to another tab and Kibana is not shown
+    document.addEventListener('visibilitychange', this.onVisiblityChangeListener);
+
+    // Count any clicks and assign it to the current app
+    window.addEventListener('click', this.onClickListener);
```

Comment:
For the `View` clicks, it basically means how many clicks on the screen have been made while this appView is active. tracking the clicks on a component is a more specific use case we can add in the future (`clicks_in_view` or something). 

As for minutes on screen, if we need a more sophosticated use case to track if the tracked component is in the screen viewport or not, we can use the providedAPI to stop tracking the view whenever it is out of focus. This is also a more specific use case we can add as a helper component if it was requested.

Reply:
I'd like to know if the proposed behaviour in this PR matches the expectations of the end-users of the data (@alexfrancoeur):

**The proposed behaviour in this PR:** when a view (flyout, tab, etc.) is rendered in the browser it counts the clicks (inside and outside the component) and minutes on-screen (even when outside of the scrolled viewport). Similarly to the current behaviour in the higher application level, it stops counting when the Browser's tab loses focus (either the browser's tab has been changed or the window has been minimized or sent to the background).

Does this match the expectations?

cc'ing @jasonrhodes because he's shown interest in this feature in the past.

Reply:
Thanks for the tag (and the direct slack pings @Bamieh 😉 ), I appreciate the ping for clarification of the request.

My expectation of this feature was that it behaves just like application usage (clicks = clicks into the application, minutes = time spent in application). So if I were to translate, a click, is a click into a view. Minutes on screen is the amount of minutes a user spends in that view. Here are some examples:

**Stack Management > Index Lifecycle Policies**
* This view is under management, so currently not captured
* Expectations: Clicks would report back clicks into `Index Lifecycle Policies`, minutes on screen would report back the minutes spent in `Index Lifecycle Policies`. When you leave this view, and transition to another view in management, we no longer capture minutes

**Metrics > Metric Explorer**
* This view is under the Metrics app as a tab, so currently not captured
* Expectations: Clicks would report back clicks into `Metrics Explorer` tab, minutes on screen would report back the minutes spent in `Metrics Explorer` tab. When you leave this tab, and transition to another view in Metrics, we no longer capture minutes

From the description in @afharo's proposed behavior summary, I do not believe this feature meets the requirements - but I may be misunderstanding. With the examples above, do you see the PR behavior capturing the expected metrics?

Reply:
Thanks @alexfrancoeur !

To summarize:
- minutes in view = minutes when view is active (user has kibana window in focus) and the view is open.
- clicks in a `view` will be changed to track only clicks within that view.


The minutes in view part is already covered. The clicks part needs to be fixed :+1:

Reply:
I've quickly demo'ed the implementation for alex to make sure i understood the requirements correctly.

I've also experimented with tracking both the advanced settings section and the telemetry flyouts to make sure the `click` tracking works as expected even for nested views.

The code implementation for tracking the section looks like this:

![image](https://user-images.githubusercontent.com/6191849/102278536-ac814600-3f32-11eb-970f-db443866ff36.png)

The usage data reported for the views looks like this:

<details>
<summary>Usage payload</summary>

"management": {
    "appId": "management",
    "viewId": "main",
    "clicks_total": 171,
    "clicks_7_days": 171,
    "clicks_30_days": 171,
    "clicks_90_days": 171,
    "minutes_on_screen_total": 7.698766666666668,
    "minutes_on_screen_7_days": 7.698766666666668,
    "minutes_on_screen_30_days": 7.698766666666668,
    "minutes_on_screen_90_days": 7.698766666666668,
    "views": [
      {
        "appId": "management",
        "viewId": "optInExampleFlyout",
        "clicks_total": 27,
        "clicks_7_days": 27,
        "clicks_30_days": 27,
        "clicks_90_days": 27,
        "minutes_on_screen_total": 5.239366666666666,
        "minutes_on_screen_7_days": 5.239366666666666,
        "minutes_on_screen_30_days": 5.239366666666666,
        "minutes_on_screen_90_days": 5.239366666666666
      },
      {
        "appId": "management",
        "viewId": "optInSecurityExampleFlyout",
        "clicks_total": 0,
        "clicks_7_days": 0,
        "clicks_30_days": 0,
        "clicks_90_days": 0,
        "minutes_on_screen_total": 0.06458333333333334,
        "minutes_on_screen_7_days": 0.06458333333333334,
        "minutes_on_screen_30_days": 0.06458333333333334,
        "minutes_on_screen_90_days": 0.06458333333333334
      },
      {
        "appId": "management",
        "viewId": "AdvancedSettings",
        "clicks_total": 11,
        "clicks_7_days": 11,
        "clicks_30_days": 11,
        "clicks_90_days": 11,
        "minutes_on_screen_total": 0.9410833333333334,
        "minutes_on_screen_7_days": 0.9410833333333334,
        "minutes_on_screen_30_days": 0.9410833333333334,
        "minutes_on_screen_90_days": 0.9410833333333334
      }
    ]
  }
}

</details>


As agreed: the `views` will only count the clicks inside of the tracked component.

As for screen time it follows the exact current implementation for the `main` view: If the user is tabbed out, changes applications or closes the component then the tracking will be paused.


In the future, we might want to support advanced use cases where we want to pause when we are outside the viewport but this can be a future enhancement *when* we have a use case for it and not for all components (under a boolean flag to enable it in specific cases).

Reply:
Thank you for the efforts, demo, and details in this thread. I love this new behaviour. I think the way we count the clicks now feels more intuitive from the analyst POV 🧡